src/vctrs.h

```diff
@@ -290,6 +280,20 @@ SEXP vec_as_names(SEXP names, enum name_repair_arg type, bool quiet);
 bool is_unique_names(SEXP names);
 SEXP vec_as_unique_names(SEXP names, bool quiet);
 
+// Character translation ----------------------------------------
+
+// `sxpinfo.gp` is exposed through `LEVELS()`, and CHARSXP info is
+// extractable through a bitwise and. This copies the unexposed
+// interface in Defn.h
+#define BYTES_MASK 2
```

Comment:
Do we have evidence that other R implementations use the same protocol for encoding? Good evidence would be that it's used in Rcpp, for example.

Reply:
Have you checked that it's worth it to use these instead of `getCharCE()`? We need to go through `LEVELS()` anyway, which is a function call because of the memory barrier, so it's not clear a priori we gain much. I'd prefer not relying on unexported API. (This should be checked with compiler optimisations enabled, maybe the branched mask inspection is optimised.)

Reply:
I got the idea from data.table actually https://github.com/Rdatatable/data.table/blob/4d355d65353eabfb2a5b58549d96b66e2266a951/src/data.table.h#L19

I'll give it a test

Reply:
Hmm, I think you might be right. I swear I thought the performance difference was more prominent, but it doesn't seem to matter _too_ much. I'll remove in favor of `Rf_getCharCE()`. The good news is that it should do exactly the same thing as what I was doing based on [its implementation](https://github.com/wch/r-source/blob/0e291d10c6631984f149320b6b30c05ce7da79ae/src/main/sysutils.c#L779).

`vec_equal()` benchmark - I switched out `CHAR_ENC_TYPE()` for `Rf_getCharCE()` [here](). This should be a "worst case scenario" because basically all of the strings are different, so that encoding check is run for every element. This would also be the worst case scenario for `Rf_getCharCE()` in general since returning `CE_NATIVE` is the last thing it does, after determining it isnt utf8/latin1/bytes.

```r
library(vctrs)
set.seed(123)

x <- stringi::stri_rand_strings(1000000, 10)
x_rev <- rev(x)

# With CHAR_ENC_TYPE()
bench::mark(
  vec_equal(x, x_rev),
  iterations = 2000
)
#> # A tibble: 1 x 6
#>   expression               min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>          <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_equal(x, x_rev)   10.4ms   10.8ms      91.8    3.87MB     9.92

# with Rf_getCharCE()
bench::mark(
  vec_equal(x, x_rev),
  iterations = 2000
)
#> # A tibble: 1 x 6
#>   expression               min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr>          <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 vec_equal(x, x_rev)   10.7ms   11.2ms      88.7    3.87MB     9.58
```

`vec_match()` benchmark - I switched out `CHAR_ENC_TYPE()` for `Rf_getCharCE()` with the most important spot being the loop [here](https://github.com/r-lib/vctrs/pull/565/files#diff-853c27e1af01a8b6f3bd37411c2bd81bR15) which is run on both `needles` and `haystack`. I think the difference is small enough to just use the interface supplied by R directly.

```r
library(vctrs)
set.seed(123)

table_unknown <- stringi::stri_rand_strings(500000, 10)
x_unknown <- sample(table_unknown, 250000, replace = TRUE)

# with CHAR_ENC_TYPE()
bench::mark(
  vec_match(x_unknown, table_unknown),
  iterations = 2000
)
#> # A tibble: 1 x 6
#>   expression                             min median `itr/sec` mem_alloc
#>   <bch:expr>                          <bch:> <bch:>     <dbl> <bch:byt>
#> 1 vec_match(x_unknown, table_unknown) 38.4ms 56.8ms      18.2    7.82MB
#> # … with 1 more variable: `gc/sec` <dbl>

# with Rf_getCharCE()
bench::mark(
  vec_match(x_unknown, table_unknown),
  iterations = 2000
)
#> # A tibble: 1 x 6
#>   expression                             min median `itr/sec` mem_alloc
#>   <bch:expr>                          <bch:> <bch:>     <dbl> <bch:byt>
#> 1 vec_match(x_unknown, table_unknown) 54.9ms 59.4ms      16.8    7.82MB
#> # … with 1 more variable: `gc/sec` <dbl>
```